### PR TITLE
Added support for snoop file

### PR DESCRIPTION
**Pull request by miketruman**
This PR will add support for reading Snoop files with PcapPlusPlus.

### CODE REVIEW

File: Pcap++/src/PcapFileDevice.cpp [hunk 1]
```diff
@@ -114,6 +118,93 @@ int IFileReaderDevice::getNextPackets(RawPacketVector& packetVec, int numOfPacke
...
+}
+
+bool SnoopFileReaderDevice::open()
+{
+	m_NumOfPacketsRead = 0;
+	m_NumOfPacketsNotParsed = 0;
+
+	m_snoopFile.open(m_FileName.c_str(), std::ifstream::binary);
+	snoop_file_header_t snoop_file_header;
+	m_snoopFile.read((char*)&snoop_file_header, sizeof(snoop_file_header_t));
```
Comment (seladb) [inline Pcap++/src/PcapFileDevice.cpp:137]: Here we should probably check if enough bytes were read:
```cpp
m_snoopFile.read((char*)&snoop_file_header, sizeof(snoop_file_header_t));
if (!m_snoopFile)
{
	PCPP_LOG_ERROR("Cannot read snoop file header for '" << m_FileName << "');
	m_snoopFile.close();
	return false;
}
```

File: Pcap++/src/PcapFileDevice.cpp [hunk 2]
```diff
@@ -114,6 +118,93 @@ int IFileReaderDevice::getNextPackets(RawPacketVector& packetVec, int numOfPacke
...
+		LINKTYPE_IEEE802_5,	/* IEEE 802.5 */
+		LINKTYPE_NULL,		/* IEEE 802.6 Metro Net */
+		LINKTYPE_ETHERNET,	/* Ethernet */
+		LINKTYPE_C_HDLC,	/* HDLC */
+		LINKTYPE_NULL,		/* Character Synchronous, e.g. bisync */
+		LINKTYPE_NULL,		/* IBM Channel-to-Channel */
+		LINKTYPE_FDDI		/* FDDI */
+	};
+
+	m_PcapLinkLayerType = snoop_encap[be32toh(snoop_file_header.datalink_type)];
```
Comment (seladb) [inline Pcap++/src/PcapFileDevice.cpp:155]: We can check that the datalink type is in range:
```cpp
if (snoop_file_header.datalink_type > 9)
{
	PCPP_LOG_ERROR("Cannot read data link type for '" << m_FileName << "');
	m_snoopFile.close();
	return false;
}
```

File: Pcap++/header/PcapFileDevice.h [hunk 3]
```diff
@@ -153,6 +154,86 @@ namespace pcpp
...
+		 * or if reached end-of-file
+		 */
+		bool getNextPacket(RawPacket& rawPacket);
+
+		/**
+		 * Open the file name which path was specified in the constructor in a read-only mode
+		 * @return True if file was opened successfully or if file is already opened. False if opening the file failed for some reason (for example:
+		 * file path does not exist)
+		 */
+		bool open();
```
Comment (seladb) [inline Pcap++/header/PcapFileDevice.h:228]: We should probably add a corresponding `close()` method?

File: Pcap++/src/PcapFileDevice.cpp [hunk 4]
```diff
@@ -114,6 +118,93 @@ int IFileReaderDevice::getNextPackets(RawPacketVector& packetVec, int numOfPacke
...
+        }
+        size_t packetSize = be32toh(snoop_packet_header.included_length);
+        if(packetSize > 15000) {
+            return false;
+        }
+	char* pMyPacketData = new char[packetSize];
+        m_snoopFile.read(pMyPacketData, packetSize);
+        if(!m_snoopFile) {
+            return false;
+        }
```
Comment (seladb) [inline Pcap++/src/PcapFileDevice.cpp:191]: Fix indentation (tabs instead of spaces)

File: Pcap++/src/PcapFileDevice.cpp [hunk 5]
```diff
@@ -114,6 +118,93 @@ int IFileReaderDevice::getNextPackets(RawPacketVector& packetVec, int numOfPacke
...
+        snoop_packet_header_t snoop_packet_header;
+        m_snoopFile.read((char*)&snoop_packet_header, sizeof(snoop_packet_header_t));
+        if(!m_snoopFile) {
+            return false;
+        }
+        size_t packetSize = be32toh(snoop_packet_header.included_length);
+        if(packetSize > 15000) {
+            return false;
+        }
+	char* pMyPacketData = new char[packetSize];
```
Comment (seladb) [inline Pcap++/src/PcapFileDevice.cpp:187]: We can probably rename this variable to `packetData` 🤔

File: Pcap++/src/PcapFileDevice.cpp [hunk 6]
```diff
@@ -114,6 +118,93 @@ int IFileReaderDevice::getNextPackets(RawPacketVector& packetVec, int numOfPacke
...
+		PCPP_LOG_ERROR("File device '" << m_FileName << "' not opened");
+		return false;
+	}
+        snoop_packet_header_t snoop_packet_header;
+        m_snoopFile.read((char*)&snoop_packet_header, sizeof(snoop_packet_header_t));
+        if(!m_snoopFile) {
+            return false;
+        }
+        size_t packetSize = be32toh(snoop_packet_header.included_length);
+        if(packetSize > 15000) {
```
Comment (seladb) [inline Pcap++/src/PcapFileDevice.cpp:184]: just curious: why is the 15000-byte limitation (and not some other number)?
I don't think we should change it, just curious why you chose 15000...

File: Tests/Pcap++Test/Tests/FileTests.cpp [hunk 7]
```diff
@@ -751,3 +751,40 @@ PTF_TEST_CASE(TestPcapFileReadLinkTypeIPv4)
...
+{
+	pcpp::SnoopFileReaderDevice readerDev(EXAMPLE_SOLARIS_SNOOP);
+	PTF_ASSERT_TRUE(readerDev.open());
+	pcpp::RawPacket rawPacket;
+	int packetCount = 0;
+	int ethCount = 0;
+	int ipCount = 0;
+	int tcpCount = 0;
+	int udpCount = 0;
+	while (readerDev.getNextPacket(rawPacket))
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/FileTests.cpp:764]: We should probably also check the timestamp of at least one packet

File: Tests/Pcap++Test/Tests/FileTests.cpp [hunk 8]
```diff
@@ -751,3 +751,40 @@ PTF_TEST_CASE(TestPcapFileReadLinkTypeIPv4)
...
+			udpCount++;
+	}
+
+	pcpp::IPcapDevice::PcapStats readerStatistics;
+
+	readerDev.getStatistics(readerStatistics);
+	PTF_ASSERT_EQUAL((uint32_t)readerStatistics.packetsRecv, 250);
+	PTF_ASSERT_EQUAL((uint32_t)readerStatistics.packetsDrop, 0);
+
+	PTF_ASSERT_EQUAL(ethCount, 142);
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/FileTests.cpp:784]: This file contains 250 packets, I wonder why Eth count isn't 250? 😳

File: Tests/Pcap++Test/main.cpp [hunk 9]
```diff
@@ -209,6 +209,7 @@ int main(int argc, char* argv[])
 	PTF_RUN_TEST(TestPcapNgFileReadWriteAdv, "no_network;pcap;pcapng");
 	PTF_RUN_TEST(TestPcapFileReadLinkTypeIPv6, "no_network;pcap");
 	PTF_RUN_TEST(TestPcapFileReadLinkTypeIPv4, "no_network;pcap");
+	PTF_RUN_TEST(TestSolarisSnoopFileRead, "no_network;pcap");
```
Comment (seladb) [inline Tests/Pcap++Test/main.cpp:212]: we can add a `snoop` tag: `"no_network;pcap;snoop"`

File: Examples/PcapPrinter/main.cpp [hunk 10]
```diff
@@ -82,6 +82,8 @@ std::string linkLayerToString(pcpp::LinkLayerType linkLayer)
 
 	if (linkLayer == pcpp::LINKTYPE_ETHERNET)
 		return "Ethernet";
+	if (linkLayer == pcpp::LINKTYPE_IEEE802_5)
+		return "LINKTYPE_IEEE802_5";
```
Comment (seladb) [inline Examples/PcapPrinter/main.cpp:86]: A more readable string might be: `IEEE 802.5 Token Ring`

File: Examples/PcapPrinter/main.cpp [hunk 11]
```diff
@@ -116,6 +118,12 @@ std::string printFileSummary(pcpp::IFileReaderDevice* reader)
 		pcpp::LinkLayerType linkLayer = pcapReader->getLinkLayerType();
 		stream << "   Link layer type: " << linkLayerToString(linkLayer) << std::endl;
 	}
+	else if (dynamic_cast<pcpp::SnoopFileReaderDevice*>(reader) != NULL)
```
Comment (seladb) [inline Examples/PcapPrinter/main.cpp:121]: We can probably add a test for a snoop file in [`test_pcapprinter.py`](https://github.com/seladb/PcapPlusPlus/blob/master/Tests/ExamplesTest/tests/test_pcapprinter.py)

File: Pcap++/src/PcapFileDevice.cpp (hunk 1 above)
Comment (miketruman) [reply Pcap++/src/PcapFileDevice.cpp:144]: Resolved in 85cf49c

File: Pcap++/src/PcapFileDevice.cpp (hunk 2 above)
Comment (miketruman) [reply Pcap++/src/PcapFileDevice.cpp]: Resolved in 2e24bdf

File: Pcap++/header/PcapFileDevice.h (hunk 3 above)
Comment (miketruman) [reply Pcap++/header/PcapFileDevice.h:228]: Resolved in 48bcd4c

File: Pcap++/src/PcapFileDevice.cpp (hunk 4 above)
Comment (miketruman) [reply Pcap++/src/PcapFileDevice.cpp]: Resolved in f37a8c2

File: Pcap++/src/PcapFileDevice.cpp (hunk 6 above)
Comment (miketruman) [reply Pcap++/src/PcapFileDevice.cpp]: To be honest, I just added something big - I would assume the largest possible packet would be 9000 Jumbo frame

File: Tests/Pcap++Test/Tests/FileTests.cpp (hunk 7 above)
Comment (miketruman) [reply Tests/Pcap++Test/Tests/FileTests.cpp:766]: Resolved in 1e32b15

File: Tests/Pcap++Test/Tests/FileTests.cpp (hunk 8 above)
Comment (miketruman) [reply Tests/Pcap++Test/Tests/FileTests.cpp:789]: The remaining packets are EthernetDot3.  I extended the snoop testcase in commit 1e32b15 with check for 108 EthernetDot3 frames.

File: Examples/PcapPrinter/main.cpp (hunk 11 above)
Comment (miketruman) [reply Examples/PcapPrinter/main.cpp:121]: Added testcase in 27c7d04

File: Examples/PcapPrinter/main.cpp (hunk 10 above)
Comment (miketruman) [reply Examples/PcapPrinter/main.cpp]: Resolved in 27c7d04

File: Tests/Pcap++Test/main.cpp (hunk 9 above)
Comment (miketruman) [reply Tests/Pcap++Test/main.cpp]: Added in 27c7d04

**Review by seladb — APPROVED** [review]

**Pull request merged**